LandSandBoat/server
# Limbus SE Apollyon Area

**_I affirm:_**

## What does this pull request do?

On a very high level, this refactors the SE Apollyon Limbus area and fixes some missing mechanics within the area. In order to achieve this though there have been many changes necessary. I managed to put many changes into other PRs that have already been merged but this one is still quite beefy. I will do my best to breakdown the changes; however, due to the large number of changes I fully anticipate this PR requiring some iteration.

### [Allow mobs to be triggered while they are "normal" status. Right now …](https://github.com/LandSandBoat/server/commit/68a0c38e2d5404c400e4d46f52d25b488b6944e6)

This change is simply to make it possible to interact with mobs. My understanding is that the client will only send a trigger event to the server when the entity has a normal status which is what gives them a green name. Unfortunately making it possible to interact with a mob via a "onTrigger" in the mob's file would require changing more. We may or may not want to do this?

### [Add support for setting up superlinking within a battlefield group](https://github.com/LandSandBoat/server/commit/67b6e0d74c9b1f844a3bac67251417eead9187ac)

I didn't realize until after the fact that this change isn't needed until a future Limbus area but here it is now!

### [Fixing critical issue with CBaseEntity::IsTargetable()](https://github.com/LandSandBoat/server/commit/c9280a5b55dbbfc6dcf307900f5fc7cbeb32bc39)

I needed a way to prevent a mob that has had its status changed to NORMAL from being hit with AOE spells and discovered that this was coded improperly.

### [Support for recovery crate. It is actually a MOB that looks like an NPC…](https://github.com/LandSandBoat/server/commit/aba7b881d5d2bc4e9352fb1a7f00ecbd7c47d2f1)

It is so beautiful!

### Refactor of SE Apollyon

One of my main objectives with these changes is to reduce duplicated code and bring as much Limbus area related content into a single file as I could. I introduced `LimbusArea` based off of `Mission` that assists with setting everything up. In addition it will create the battlefield object which contains a significant amount of duplicated code that exists in all Limbus area files.

All of this new code exists alongside the majority of the old Limbus code that is still being used for the other areas. Once all areas have been converted over to this new implementation then I will cleanup the old code.

### Regarding Loot

Loot right now is very much WIP as I'm still figuring out how best to balance the relic materials. I'm testing the waters of using the distribution method discussed over at https://github.com/LandSandBoat/server/discussions/2767. My current plan is to finish the main Apollyon areas and then submit a separate PR specifically for reworking the loot in those areas.

## Steps to test these changes

```
!zone 38
!addkeyitem cosmo_cleanse
!addkeyitem black_card
```

Enter into SE Apollyon and progress through the area. The following captures were used and can be referenced to verify mechanics are working properly.

https://discord.com/channels/443544205206355968/653844344088035328/704860303141371954
https://discord.com/channels/443544205206355968/653844344088035328/704877361757159494
https://discord.com/channels/443544205206355968/653844344088035328/668861099029954562

Other useful links:
https://ffxiclopedia.fandom.com/wiki/SE_Apollyon
https://www.bg-wiki.com/ffxi/South_East_Apollyon
